boa_gc/src/internals/gc_box.rs

```diff
@@ -0,0 +1,155 @@
+use crate::Trace;
+use std::cell::Cell;
+use std::ptr::{self, NonNull};
+
+// Age and Weak Flags
+const MARK_MASK: usize = 1 << (usize::BITS - 2);
+const WEAK_MASK: usize = 1 << (usize::BITS - 1);
+const ROOTS_MASK: usize = !(MARK_MASK | WEAK_MASK);
+const ROOTS_MAX: usize = ROOTS_MASK;
+
+pub(crate) struct GcBoxHeader {
+    roots: Cell<usize>,
+    pub(crate) next: Cell<Option<NonNull<GcBox<dyn Trace>>>>,
+}
+
+impl GcBoxHeader {
+    #[inline]
+    pub(crate) fn new() -> Self {
+        GcBoxHeader {
+            roots: Cell::new(1),
+            next: Cell::new(None),
+        }
+    }
+
+    #[inline]
+    pub(crate) fn new_weak() -> Self {
+        // Set weak_flag
+        GcBoxHeader {
+            roots: Cell::new(WEAK_MASK | 1),
+            next: Cell::new(None),
+        }
+    }
+
+    #[inline]
+    pub(crate) fn roots(&self) -> usize {
+        self.roots.get() & ROOTS_MASK
+    }
+
+    #[inline]
+    pub(crate) fn inc_roots(&self) {
+        let roots = self.roots.get();
+
+        if (roots & ROOTS_MASK) < ROOTS_MAX {
+            self.roots.set(roots + 1);
+        } else {
+            // TODO: implement a better way to handle root overload
+            panic!("roots counter overflow");
+        }
+    }
+
+    #[inline]
+    pub(crate) fn dec_roots(&self) {
+        // Underflow check as a stop gap for current issue when dropping
+        if self.roots.get() > 0 {
+            self.roots.set(self.roots.get() - 1);
+        }
+    }
+
+    #[inline]
+    pub(crate) fn is_marked(&self) -> bool {
+        self.roots.get() & MARK_MASK != 0
+    }
+
+    #[inline]
+    pub(crate) fn mark(&self) {
+        self.roots.set(self.roots.get() | MARK_MASK);
+    }
+
+    #[inline]
+    pub(crate) fn unmark(&self) {
+        self.roots.set(self.roots.get() & !MARK_MASK);
+    }
+
+    #[inline]
+    pub(crate) fn is_ephemeron(&self) -> bool {
+        self.roots.get() & WEAK_MASK != 0
+    }
+}
+
+// NOTE: [repr(C)] is most likely unneeded here, but will keep it for now
+/// A garbage collected allocation.
+#[repr(C)]
+pub(crate) struct GcBox<T: Trace + ?Sized + 'static> {
+    pub(crate) header: GcBoxHeader,
+    pub(crate) value: T,
+}
+
+impl<T: Trace> GcBox<T> {
+    pub(crate) fn new(value: T) -> Self {
+        GcBox {
```

Comment:
```suggestion
        Self {
```
This can in fact be done in many places, and can make renames easier at some point.